### PR TITLE
SALTO-1739: Add support for "retry-after" header

### PR DESCRIPTION
**Pull request by alonstern**
Added support for waiting the time that was returned in the "retry-after" header in the response before retrying a request in the simple adapter HTTP client

---
_Release Notes_: 
None

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/adapter-components/src/client/http_connection.ts [hunk 1]
```diff
@@ -70,14 +71,26 @@ export type ConnectionCreator<TCredentials> = (
...
   retries: retryOptions.maxAttempts,
   retryDelay: (retryCount, err) => {
+    // Although the standard is 'Retry-After' is seems that some servers
+    // returns 'retry-after' so just in case we lowercase the headers
+    const retryAfterHeaderValue = _.mapKeys(
+      err.response?.headers ?? {},
+      (_val, key) => key.toLowerCase()
+    )['retry-after']
+    const retryDelay = retryAfterHeaderValue !== undefined
+      ? parseInt(retryAfterHeaderValue, 10) * 1000
```
Comment (ori-moisis) [inline packages/adapter-components/src/client/http_connection.ts:81]: `parseInt` could return `NaN` if the value in the header is malformed, in which case I assume something down the line will fail in an unexpected way
we should cover that case

File: packages/adapter-components/test/client/http_connection.test.ts [hunk 2]
```diff
@@ -70,4 +71,44 @@ describe('client_http_connection', () => {
...
+          },
+        },
+        code: 'code',
+        config: {
+          url: 'url',
+        },
+      } as AxiosError)).toBe(10000)
+    })
+
+    it('should use the input detail when retry-after header when available', () => {
```
Comment (ori-moisis) [inline packages/adapter-components/test/client/http_connection.test.ts:102]: I don't understand what this sentence is supposed to say

File: packages/adapter-components/test/client/http_connection.test.ts [hunk 3]
```diff
@@ -70,4 +71,44 @@ describe('client_http_connection', () => {
...
+    })
+    it('should retry error code 429', () => {
+      expect(retryOptions.retryCondition?.({
+        response: {
+          status: 429,
+        },
+      } as AxiosError)).toBeTruthy()
+    })
+
+    it('should use the retry-after header when available', () => {
```
Comment (ori-moisis) [inline packages/adapter-components/test/client/http_connection.test.ts:88]: the code has specific handling for lower case headers, we should have a test for that.

also, I think we should have a test for "strange" cases like getting 429 without the header and the header value not being an integer, we should make sure we don't completely fail in such cases (probably best to fall back to the configured delay)

File: packages/adapter-components/test/client/http_connection.test.ts [hunk 4]
```diff
@@ -70,4 +71,71 @@ describe('client_http_connection', () => {
...
+        response: {
+          headers: {
+            'retry-after': '10',
+          },
+        },
+        code: 'code',
+        config: {
+          url: 'url',
+        },
+      } as AxiosError)).toBe(10000)
```
Comment (ori-moisis) [inline packages/adapter-components/test/client/http_connection.test.ts:111]: why is this here twice? is there a difference I am missing?

File: packages/adapter-components/test/client/http_connection.test.ts [hunk 5]
```diff
@@ -70,4 +71,71 @@ describe('client_http_connection', () => {
...
+        response: {
+          headers: {
+            'Retry-After': 'invalid',
+          },
+        },
+        code: 'code',
+        config: {
+          url: 'url',
+        },
+      } as AxiosError)).toBe(100)
```
Comment (ori-moisis) [inline packages/adapter-components/test/client/http_connection.test.ts:138]: return status here is not 429, doesn't that make the test invalid?

File: packages/adapter-components/test/client/http_connection.test.ts (hunk 4 above)
Comment (alonstern) [reply packages/adapter-components/test/client/http_connection.test.ts:111]: The case of the retry after key is different

File: packages/adapter-components/test/client/http_connection.test.ts (hunk 5 above)
Comment (alonstern) [reply packages/adapter-components/test/client/http_connection.test.ts:138]: the status code is only checked in the `retryCondition` function so it does not have an effect on this.
This is because the `Retry-After` can also be sent with other statuses, e.g. 503 (which is covered by the `isNetworkOrIdempotentRequestError` in the retryCodition function)

Comment (ori-moisis) [reply packages/adapter-components/test/client/http_connection.test.ts:138]: got it

**Review by ori-moisis — APPROVED** [review]
